File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int>ans;
    void dfs(TreeNode* root,int dep){
        if(!root)return;
        if(ans.size() == dep)ans.push_back(root->val);
        else ans[dep] = max(ans[dep],root->val);
        dfs(root->left,dep+1);
        dfs(root->right,dep+1);
    }
    vector<int> largestValues(TreeNode* root) {
        if(!root)return {};
        queue<TreeNode*>q;
        vector<int>ans;
        q.push(root);
        while(q.size()){
            int sz = q.size();//2
            int mx = INT_MIN; 
            while(sz--){
                auto node = q.front();q.pop();
                mx=max(mx,node->val);
                if(node->left)q.push(node->left);
                if(node->right)q.push(node->right);
            }
            ans.push_back(mx);
        }
        return ans;
        
        // Solution #2
        dfs(root,0);
        return ans;
    }
};

File: 0515-find-largest-value-in-each-tree-row/0515-find-largest-value-in-each-tree-row.py
from collections import deque
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def dfs(self, root: Optional[TreeNode], dep: int):
        if not root:
            return
        if len(self.ans) == dep:
            self.ans.append(root.val)
        else:
            self.ans[dep] = max(self.ans[dep], root.val)
        self.dfs(root.left, dep + 1)
        self.dfs(root.right, dep + 1)

    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        # Solution #1: BFS
        q = deque([root])
        ans = []
        while q:
            sz = len(q)
            mx = float('-inf')
            for _ in range(sz):
                node = q.popleft()
                mx = max(mx, node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            ans.append(mx)

        # Solution #2: DFS
        self.ans = []
        self.dfs(root, 0)
        return self.ans
